src/core/service.c

```diff
@@ -2144,6 +2145,54 @@ _pure_ static bool service_can_reload(Unit *u) {
         return !!s->exec_command[SERVICE_EXEC_RELOAD];
 }
 
+static int service_exec_command_index(Unit *u, ServiceExecCommand id, bool control) {
+        Service *s = SERVICE(u);
+        int index = 0;
+        ExecCommand *current, *first, *c;
+
+        assert(s);
+
+        first = s->exec_command[id];
+        current = control ? s->control_command : s->main_command;
+
+        /* Figure out where we are in the list be walking back to the beginning */
+        for (c = current; c != first; c = c->command_prev)
+                index++;
+
+        return index;
+}
+
+static int service_serialize_exec_command(Unit *u, FILE *f, bool control) {
+        Service *s = SERVICE(u);
+        ServiceExecCommand id;
+        ExecCommand *c;
+        int index;
+        const char *type;
+        _cleanup_free_ char *args = NULL;
+
+        assert(s);
+        assert(f);
+
+        if (control) {
+                type = "control";
+                c = s->control_command;
+                id = s->control_command_id;
+        } else {
+                type = "main";
+                c = s->main_command;
+                id = SERVICE_EXEC_START;
+        }
+
+        assert(c);
+
+        index = service_exec_command_index(u, id, control);
+        args = strv_join(c->argv, " ");
+
+        fprintf(f, "%s-command=%s %d %s %s\n", type, service_exec_command_to_string(id), index, c->path, args);
```

Comment:
hmm, so we might want to protect ourselves here from command lines that contain a newline. Those would result in borked serialization. I think it would be best to use something smarter than strv_join() for serializing the command line: build your own loop, and concatenate the arguments manually, after processing each individually through cescape(), escaping everything weird, including newlines and spaces.